main.go

```diff
@@ -78,55 +91,55 @@ func createConfigForHealthMonitor(restConfig *rest.Config) *rest.Config {
 	return &restConfigCopy
 }
 
-func newCustomTransportWrapper(lb LoadBalancer) func(http.RoundTripper) http.RoundTripper {
-	return func(rt http.RoundTripper) http.RoundTripper { return &customTransport{baseRT: rt, lb: lb} }
+func newCustomTransportWrapper(lb LoadBalancer, cbGetter func(server string) CircuitBreaker) func(http.RoundTripper) http.RoundTripper {
+	return func(rt http.RoundTripper) http.RoundTripper { return &customTransport{baseRT: rt, lb: lb, cbGetter: cbGetter} }
 }
 
 type customTransport struct {
 	baseRT http.RoundTripper
 	lb     LoadBalancer
+	cbGetter func(server string) CircuitBreaker
 }
 
 func (t *customTransport) RoundTrip(r *http.Request) (*http.Response, error) {
 	fmt.Println(fmt.Sprintf("customTransport: RoundTrip: received a request for Method = %q to Host = %q, URL = %q", r.Method, r.Host, r.URL))
 
+	// TODO: filter on specific path (for example only SAR requests)
 	// TODO: if there are no healthy targets we could use the service IP - why not ?
 	server, err := t.lb.NextServer()
 	if err != nil {
 		return nil, err
 	}
+	seenServersSet := sets.NewString(server)
 
-	// I don't like modifying the original request
-	r.Host = server
-	r.URL.Host = server
-
-	fmt.Println(fmt.Sprintf("customTransport: RoundTrip: The new Host is = %q", r.Host))
-
-	// TODO: needs proper implementation
-	rt := getCircuitBreakerForEndpoint(server, t.baseRT)
-	return rt.RoundTrip(r)
-}
-
-// getCircuitBreakerForEnpoint could map an endpoint (URL) to a circuit breaker aware endpoint
-// the mapper could have a cache (LRU) to store the mapping
-func getCircuitBreakerForEndpoint(ep string, delegate http.RoundTripper) http.RoundTripper {
-	return &cb{delegate: delegate}
-}
+	for {
+		// I don't like modifying the original request
+		r.Host = server
+		r.URL.Host = server
+
+		fmt.Println(fmt.Sprintf("customTransport: RoundTrip: The new Host is = %q", r.Host))
+
+		crt := t.cbGetter(server)
+		rsp, err := crt.RoundTripWithDelegate(r, t.baseRT)
+		if rsp == nil && err != nil && crt.FailedBecauseCircuitBreakerIsOpen(err) {
```

Comment:
when can we get err!=nil ? Is there a case with rsp!=nil and err!=nil we want not to retry?

Reply:
> Is there a case with rsp!=nil and err!=nil we want not to retry?

yes, retry is done by client-go

`rsp == nil && err != nil && crt.FailedBecauseCircuitBreakerIsOpen(err)`

is a special case that tells us that the `cb` was `Open`